File: backend/src/routes/identity.routes.js
const express = require('express');
const { requireAuth } = require('../middleware/auth.middleware');
const identityService = require('../services/identity.service');
const trustPlusService = require('../services/trust-plus.service');
const { checkAndRecord } = require('../utils/rateLimit');
const { catchAsync } = require('../utils/helpers');
const ApiError = require('../utils/ApiError');

const router = express.Router();

const AADHAAR_RE = /^\d{12}$/;
const PAN_RE = /^[A-Z]{5}\d{4}[A-Z]$/;
const GST_RE = /^\d{2}[A-Z]{5}\d{4}[A-Z][A-Z0-9][Z][A-Z0-9]$/;
const IFSC_RE = /^[A-Z]{4}0[A-Z0-9]{6}$/;

const enforceKycRateLimit = (userId, docType) => {
  const { allowed, remaining } = checkAndRecord(`kyc:${docType}:${userId}`, 5, 86400);
  if (!allowed) {
    throw new ApiError(429, `Too many ${docType} submissions today. Try again in ${remaining}s`);
  }
};

const maybeAwardTrustPlus = async (userId) => {
  try {
    await trustPlusService.maybeAwardBonus(userId);
  } catch (err) {
    // Non-fatal
  }
};

router.post('/aadhaar/verify', requireAuth, catchAsync(async (req, res) => {
  const { aadhaar_number, doc_url } = req.body;
  if (!aadhaar_number || !AADHAAR_RE.test(aadhaar_number)) {
    throw ApiError.badRequest('Invalid Aadhaar format (12 digits)');
  }
  if (!doc_url) {
    throw ApiError.badRequest('doc_url is required');
  }

  const userId = req.user._id.toString();
  enforceKycRateLimit(userId, 'aadhaar');

  const doc = await identityService.submitDocument({
    userId,
    docType: 'aadhaar',
    docNumber: aadhaar_number,
    docUrl: doc_url,
  });

  await maybeAwardTrustPlus(userId);
  res.json(doc);
}));

router.post('/pan/verify', requireAuth, catchAsync(async (req, res) => {
  const { pan_number, doc_url } = req.body;
  if (!pan_number || !PAN_RE.test(pan_number.toUpperCase())) {
    throw ApiError.badRequest('Invalid PAN format (e.g. ABCDE1234F)');
  }
  if (!doc_url) {
    throw ApiError.badRequest('doc_url is required');
  }

  const userId = req.user._id.toString();
  enforceKycRateLimit(userId, 'pan');

  const doc = await identityService.submitDocument({
    userId,
    docType: 'pan',
    docNumber: pan_number.toUpperCase(),
    docUrl: doc_url,
  });

  await maybeAwardTrustPlus(userId);
  res.json(doc);
}));

router.post('/gst/verify', requireAuth, catchAsync(async (req, res) => {
  const { gst_number, doc_url } = req.body;
  if (!gst_number || !GST_RE.test(gst_number.toUpperCase())) {
    throw ApiError.badRequest('Invalid GST format (15 chars)');
  }
  if (!doc_url) {
    throw ApiError.badRequest('doc_url is required');
  }

  const userId = req.user._id.toString();
  enforceKycRateLimit(userId, 'gst');

  const doc = await identityService.submitDocument({
    userId,
    docType: 'gst',
    docNumber: gst_number.toUpperCase(),
    docUrl: doc_url,
  });

  await maybeAwardTrustPlus(userId);
  res.json(doc);
}));

router.post('/bank/verify', requireAuth, catchAsync(async (req, res) => {
  const { account_number, ifsc, holder_name, bank_name, doc_url } = req.body;
  if (!ifsc || !IFSC_RE.test(ifsc.toUpperCase())) {
    throw ApiError.badRequest('Invalid IFSC format');
  }
  if (!account_number || account_number.length < 6 || account_number.length > 32) {
    throw ApiError.badRequest('Account number must be between 6 and 32 characters');
  }
  if (!holder_name || holder_name.length < 2 || holder_name.length > 120) {
    throw ApiError.badRequest('Holder name must be between 2 and 120 characters');
  }
  if (!doc_url) {
    throw ApiError.badRequest('doc_url is required');
  }

  const userId = req.user._id.toString();
  enforceKycRateLimit(userId, 'bank');

  const doc = await identityService.submitDocument({
    userId,
    docType: 'bank',
    docNumber: account_number,
    docUrl: doc_url,
    additionalData: {
      ifsc: ifsc.toUpperCase(),
      account_number,
      holder_name: String(holder_name).trim(),
      bank_name: bank_name ? String(bank_name).trim() : null,
    },
  });

  await maybeAwardTrustPlus(userId);
  res.json(doc);
}));

router.get('/trust-plus/me', requireAuth, catchAsync(async (req, res) => {
  const status = await trustPlusService.computeStatus(req.user._id.toString());
  res.json(status);
}));

router.get('/me/status', requireAuth, catchAsync(async (req, res) => {
  const summary = await identityService.getStatusSummary(req.user._id.toString());
  res.json(summary);
}));

router.get('/me/docs', requireAuth, catchAsync(async (req, res) => {
  const items = await identityService.listDocuments(req.user._id.toString());
  res.json({ items });
}));

router.delete('/docs/:doc_id', requireAuth, catchAsync(async (req, res) => {
  const result = await identityService.deleteDocument(req.user._id.toString(), req.params.doc_id);
  res.json(result);
}));

module.exports = router;

File: backend/routes/identity_routes.py
"""Identity Verification routes — 4 doc types (aadhaar/pan/gst/bank).

Rate limit: 5/day/user per doc-type via utils/rate_limit.
"""
from __future__ import annotations
import re
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from middleware.auth_middleware import require_auth
from services import identity_service
from utils.rate_limit import check_and_record

router = APIRouter(prefix="/v1/kyc", tags=["identity"])

# Simple format validators. Real KYC providers do deeper checks.
_AADHAAR_RE = re.compile(r"^\d{12}$")
_PAN_RE = re.compile(r"^[A-Z]{5}\d{4}[A-Z]$")
_GST_RE = re.compile(r"^\d{2}[A-Z]{5}\d{4}[A-Z][A-Z0-9][Z][A-Z0-9]$")
_IFSC_RE = re.compile(r"^[A-Z]{4}0[A-Z0-9]{6}$")


def _rate(user_id: str, doc_type: str) -> None:
    ok, retry = check_and_record(f"kyc:{doc_type}:{user_id}", limit=5, window_seconds=86400)
    if not ok:
        raise HTTPException(429, f"Too many {doc_type} submissions today. Try again in {retry}s")


class AadhaarBody(BaseModel):
    aadhaar_number: str = Field(..., min_length=12, max_length=12)
    doc_url: str = Field(..., max_length=800)


class PanBody(BaseModel):
    pan_number: str = Field(..., min_length=10, max_length=10)
    doc_url: str = Field(..., max_length=800)


class GstBody(BaseModel):
    gst_number: str = Field(..., min_length=15, max_length=15)
    doc_url: str = Field(..., max_length=800)


class BankBody(BaseModel):
    account_number: str = Field(..., min_length=6, max_length=32)
    ifsc: str = Field(..., min_length=11, max_length=11)
    holder_name: str = Field(..., min_length=2, max_length=120)
    bank_name: str | None = Field(default=None, max_length=120)
    doc_url: str = Field(..., max_length=800)


@router.post("/aadhaar/verify")
async def verify_aadhaar(body: AadhaarBody, user=Depends(require_auth)):
    if not _AADHAAR_RE.match(body.aadhaar_number):
        raise HTTPException(400, "Invalid Aadhaar format (12 digits)")
    _rate(str(user.id), "aadhaar")
    return await identity_service.submit_document(
        user_id=str(user.id), doc_type="aadhaar",
        doc_number=body.aadhaar_number, doc_url=body.doc_url,
    )


@router.post("/pan/verify")
async def verify_pan(body: PanBody, user=Depends(require_auth)):
    if not _PAN_RE.match(body.pan_number.upper()):
        raise HTTPException(400, "Invalid PAN format (e.g. ABCDE1234F)")
    _rate(str(user.id), "pan")
    return await identity_service.submit_document(
        user_id=str(user.id), doc_type="pan",
        doc_number=body.pan_number.upper(), doc_url=body.doc_url,
    )


@router.post("/gst/verify")
async def verify_gst(body: GstBody, user=Depends(require_auth)):
    if not _GST_RE.match(body.gst_number.upper()):
        raise HTTPException(400, "Invalid GST format (15 chars)")
    _rate(str(user.id), "gst")
    return await identity_service.submit_document(
        user_id=str(user.id), doc_type="gst",
        doc_number=body.gst_number.upper(), doc_url=body.doc_url,
    )


@router.post("/bank/verify")
async def verify_bank(body: BankBody, user=Depends(require_auth)):
    if not _IFSC_RE.match(body.ifsc.upper()):
        raise HTTPException(400, "Invalid IFSC format")
    _rate(str(user.id), "bank")
    return await identity_service.submit_document(
        user_id=str(user.id), doc_type="bank",
        doc_number=body.account_number, doc_url=body.doc_url,
        additional_data={
            "ifsc": body.ifsc.upper(),
            "account_number": body.account_number,
            "holder_name": body.holder_name.strip(),
            "bank_name": (body.bank_name or "").strip() or None,
        },
    )


@router.get("/me/status")
async def kyc_me_status(user=Depends(require_auth)):
    return await identity_service.get_status_summary(str(user.id))


@router.get("/me/docs")
async def kyc_me_docs(user=Depends(require_auth)):
    return {"items": await identity_service.list_documents(str(user.id))}


@router.delete("/docs/{doc_id}")
async def delete_doc(doc_id: str, user=Depends(require_auth)):
    return await identity_service.delete_document(str(user.id), doc_id)
